hibernate-core/src/main/java/org/hibernate/query/sqm/tree/select/SqmSelectStatement.java

```diff
@@ -483,21 +484,49 @@ public SqmSelectStatement<Long> createCountQuery() {
 			return (SqmSelectStatement<Long>) copy;
 		}
 		else {
-			final JpaSelection<?> selection = queryPart.getFirstQuerySpec().getSelection();
-			if ( selection.isCompoundSelection() ) {
-				char c = 'a';
-				for ( JpaSelection<?> item : selection.getSelectionItems() ) {
-					item.alias( Character.toString( ++c ) + '_' );
-				}
-			}
-			else {
-				selection.alias( "a_" );
-			}
+			aliasSelections( queryPart.getFirstQuerySpec() );
 			final SqmSubQuery<?> subquery = new SqmSubQuery<>( copy, queryPart, null, nodeBuilder() );
 			final SqmSelectStatement<Long> query = nodeBuilder().createQuery( Long.class );
 			query.from( subquery );
 			query.select( nodeBuilder().count() );
 			return query;
 		}
 	}
+
+	private <S> void aliasSelections(SqmQuerySpec<S> querySpec) {
+		final JpaSelection<?> selection = querySpec.getSelection();
+		final List<? extends JpaSelection<?>> originalSelections = selection.isCompoundSelection() ?
```

Comment:
Can't we just walk the selections recursively and let these walking functions add regular selections to a list, along with aliasing like `c0`, `c1`, ...?

Reply:
I guess, that's basically what I'm doing iterating selections instead of walking them; only problem is what constitutes a "regular selection"? In other words, which visitations should we allow to add to the list? I think there might be a long list of cases, and we risk missing some and adding maintenance cost.

Reply:
The selections that `!isCompoundSelection()` I would say. And when `isCompoundSelection()` is `true`, you iterate the `getCompoundSelections()`.

Reply:
So you think it's enough to override `visitSelection`? But dynamic instantiation returns `false` for `isCompoundSelection()`, and inside `getCompoundSelections()` there might be dynamic instantiations, so I'd still e need the `instanceof` check.

Reply:
Indeed, I was hoping that `SqmDynamicInstantiation#isCompoundSelection` returns true. Well, then you'd have to special case `SqmDynamicInstantiation`.

Reply:
I've tried doing it with visitation instead of an `instanceof` check, relying on the return value being `null` for `visitDynamicInstantiation` (still had to cast the selection to `SqmVisitableNode`).

Note that I also switched to using `LinkedHashSet` instead of `ArrayList` since queries with multiple selections of the same path, e.g. the one in the updated test, would result in the same alias being used multiple times causing an error.

Reply:
Went back to a simple iteration implementation that should handle compound selections and dynamic instantiations correctly; I also added logic for aliasing non-simple query parts and a test for that as well.